File: tooling/check_versions.py
#!/usr/bin/env python3
"""Check that every version claim in the repository traces back to the
committed toolchain snapshot (verification/toolchains/current.json).

Checked:
  - skill frontmatter metadata pins (moonc-version / moon-version /
    moonrun-version, verified-date, verified-targets, verified-platform)
    match the snapshot exactly;
  - the visible verification note near the top of each SKILL.md body
    states the same versions;
  - every fixture carries a `verified` stamp whose component versions and
    date match the snapshot (i.e. fixtures were actually re-run against
    the toolchain the repo claims).

Usage: python3 tooling/check_versions.py
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

from validate_skills import parse_frontmatter

REPO_ROOT = Path(__file__).resolve().parent.parent
SNAPSHOT = REPO_ROOT / "verification" / "toolchains" / "current.json"
SKILLS_DIR = REPO_ROOT / "skills"
FIXTURES_DIR = REPO_ROOT / "verification" / "fixtures"

COMPONENT_KEY = {
    "moonc-version": "moonc",
    "moon-version": "moon",
    "moonrun-version": "moonrun",
}


def main() -> int:
    problems: list[str] = []
    snapshot = json.loads(SNAPSHOT.read_text())
    components = {c["name"]: c for c in snapshot["components"]}
    snap_date = snapshot["verification_date"]
    snap_targets = set(snapshot["verified_targets"])
    snap_platform = (
        f"{snapshot['platform']['os']}-{snapshot['platform']['arch']}"
    )

    for skill_dir in sorted(d for d in SKILLS_DIR.iterdir() if d.is_dir()):
        skill_md = skill_dir / "SKILL.md"
        if not skill_md.is_file():
            continue
        fm, body, _ = parse_frontmatter(skill_md.read_text())
        metadata = fm.get("metadata", {}) or {}
        label = skill_dir.name

        for meta_key, comp_name in COMPONENT_KEY.items():
            if meta_key not in metadata:
                continue
            expected = components[comp_name]["version"]
            if metadata[meta_key] != expected:
                problems.append(
                    f"{label}: metadata.{meta_key}={metadata[meta_key]!r} "
                    f"but snapshot has {expected!r}"
                )
            # The human-visible verification note must state the same pin.
            head = "\n".join(body.splitlines()[:40])
            if expected not in head:
                problems.append(
                    f"{label}: SKILL.md body head does not state verified "
                    f"{comp_name} version {expected!r}"
                )

        if metadata.get("verified-date") != snap_date:
            problems.append(
                f"{label}: metadata.verified-date={metadata.get('verified-date')!r}"
                f" != snapshot {snap_date!r}"
            )
        if metadata.get("verified-platform") != snap_platform:
            problems.append(
                f"{label}: metadata.verified-platform="
                f"{metadata.get('verified-platform')!r} != snapshot {snap_platform!r}"
            )
        declared_targets = set(
            t.strip() for t in metadata.get("verified-targets", "").split(",") if t.strip()
        )
        if not declared_targets or not declared_targets <= snap_targets:
            problems.append(
                f"{label}: metadata.verified-targets {sorted(declared_targets)} "
                f"must be a non-empty subset of snapshot {sorted(snap_targets)}"
            )

    for fixture_dir in sorted(FIXTURES_DIR.iterdir()):
        meta_file = fixture_dir / "fixture.json"
        if not meta_file.is_file():
            continue
        meta = json.loads(meta_file.read_text())
        stamp = meta.get("verified")
        if not stamp:
            problems.append(f"fixture {fixture_dir.name}: never stamped as verified")
            continue
        for comp_name, comp in components.items():
            stamped = stamp.get("components", {}).get(comp_name, "")
            # Stamps store the raw `moon version` strings; compare prefix-insensitively.
            if comp["version"] not in stamped:
                problems.append(
                    f"fixture {fixture_dir.name}: stamped {comp_name} "
                    f"{stamped!r} does not match snapshot {comp['version']!r}"
                )
        if stamp.get("date") != snap_date:
            problems.append(
                f"fixture {fixture_dir.name}: stamp date {stamp.get('date')!r} "
                f"!= snapshot {snap_date!r}"
            )

    for problem in problems:
        print(f"FAIL {problem}", file=sys.stderr)
    if not problems:
        print("version consistency: OK")
    return 1 if problems else 0


if __name__ == "__main__":
    sys.exit(main())

File: tooling/check_versions.ts
import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import { basename, join } from "node:path";
import { exitWith, isMain } from "./lib/cli.ts";
import { parseFrontmatter, stringMap } from "./lib/frontmatter.ts";
import { REPO_ROOT } from "./lib/repo.ts";

const SNAPSHOT = join(REPO_ROOT, "verification", "toolchains", "current.json");
const SKILLS_DIR = join(REPO_ROOT, "skills");
const FIXTURES_DIR = join(REPO_ROOT, "verification", "fixtures");

const COMPONENT_KEY: Record<string, string> = {
  "moonc-version": "moonc",
  "moon-version": "moon",
  "moonrun-version": "moonrun",
};

interface Component {
  name: string;
  version: string;
}

interface ToolchainSnapshot {
  verification_date: string;
  platform: { os: string; arch: string };
  components: Component[];
  verified_targets: string[];
}

interface FixtureStamp {
  date?: string;
  components?: Record<string, string>;
}

interface FixtureMetadata {
  verified?: FixtureStamp;
}

export function main(): number {
  const problems: string[] = [];
  const snapshot = JSON.parse(readFileSync(SNAPSHOT, "utf8")) as ToolchainSnapshot;
  const components = new Map(snapshot.components.map((component) => [component.name, component]));
  const snapshotDate = snapshot.verification_date;
  const snapshotTargets = new Set(snapshot.verified_targets);
  const snapshotPlatform = `${snapshot.platform.os}-${snapshot.platform.arch}`;

  for (const skillDirectory of directories(SKILLS_DIR)) {
    const skillMd = join(skillDirectory, "SKILL.md");
    if (!isFile(skillMd)) {
      continue;
    }
    const parsed = parseFrontmatter(readFileSync(skillMd, "utf8"));
    const metadata = stringMap(parsed.frontmatter, "metadata") ?? {};
    const label = basename(skillDirectory);

    if (!("moonbit-release" in metadata)) {
      continue;
    }

    for (const [metadataKey, componentName] of Object.entries(COMPONENT_KEY)) {
      if (!(metadataKey in metadata)) {
        continue;
      }
      const expected = components.get(componentName)?.version;
      if (expected === undefined) {
        throw new Error(`snapshot missing component ${componentName}`);
      }
      if (metadata[metadataKey] !== expected) {
        problems.push(
          `${label}: metadata.${metadataKey}=${repr(metadata[metadataKey])} but snapshot has ${repr(expected)}`,
        );
      }
      const head = parsed.body.split("\n").slice(0, 40).join("\n");
      if (!head.includes(expected)) {
        problems.push(
          `${label}: SKILL.md body head does not state verified ${componentName} version ${repr(expected)}`,
        );
      }
    }

    if (metadata["verified-date"] !== snapshotDate) {
      problems.push(
        `${label}: metadata.verified-date=${reprOptional(metadata["verified-date"])} != snapshot ${repr(snapshotDate)}`,
      );
    }
    if (metadata["verified-platform"] !== snapshotPlatform) {
      problems.push(
        `${label}: metadata.verified-platform=${reprOptional(metadata["verified-platform"])} != snapshot ${repr(snapshotPlatform)}`,
      );
    }
    const declaredTargets = new Set(
      (metadata["verified-targets"] ?? "")
        .split(",")
        .map((target) => target.trim())
        .filter(Boolean),
    );
    if (
      declaredTargets.size === 0 ||
      [...declaredTargets].some((target) => !snapshotTargets.has(target))
    ) {
      problems.push(
        `${label}: metadata.verified-targets ${reprList([...declaredTargets].sort())} must be a non-empty subset of snapshot ${reprList([...snapshotTargets].sort())}`,
      );
    }
  }

  for (const fixtureDirectory of directories(FIXTURES_DIR)) {
    const metadataFile = join(fixtureDirectory, "fixture.json");
    if (!isFile(metadataFile)) {
      continue;
    }
    const metadata = JSON.parse(readFileSync(metadataFile, "utf8")) as FixtureMetadata;
    const stamp = metadata.verified;
    const fixture = basename(fixtureDirectory);
    if (stamp === undefined) {
      problems.push(`fixture ${fixture}: never stamped as verified`);
      continue;
    }
    for (const component of snapshot.components) {
      const stamped = stamp.components?.[component.name] ?? "";
      if (!stamped.includes(component.version)) {
        problems.push(
          `fixture ${fixture}: stamped ${component.name} ${repr(stamped)} does not match snapshot ${repr(component.version)}`,
        );
      }
    }
    if (stamp.date !== snapshotDate) {
      problems.push(
        `fixture ${fixture}: stamp date ${reprOptional(stamp.date)} != snapshot ${repr(snapshotDate)}`,
      );
    }
  }

  for (const problem of problems) {
    console.error(`FAIL ${problem}`);
  }
  if (problems.length === 0) {
    console.log("version consistency: OK");
  }
  return problems.length > 0 ? 1 : 0;
}

function directories(parent: string): string[] {
  return readdirSync(parent, { withFileTypes: true })
    .filter((entry) => entry.isDirectory())
    .map((entry) => join(parent, entry.name))
    .sort();
}

function isFile(path: string): boolean {
  return existsSync(path) && statSync(path).isFile();
}

function repr(value: string): string {
  return `'${value.replaceAll("\\", "\\\\").replaceAll("'", "\\'")}'`;
}

function reprOptional(value: string | undefined): string {
  return value === undefined ? "None" : repr(value);
}

function reprList(values: string[]): string {
  return `[${values.map(repr).join(", ")}]`;
}

if (isMain(import.meta.url)) {
  exitWith(main());
}
